File: Proj4a/Questão 4/Quick-sort/quick-sort.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

// Função para trocar dois elementos
void swap(int* a, int* b) {
    int t = *a;
    *a = *b;
    *b = t;
}

// Função para particionar o array
int partition(int arr[], int low, int high) {
    int pivot = arr[high];
    int i = (low - 1);

    for (int j = low; j <= high - 1; j++) {
        if (arr[j] < pivot) {
            i++;
            swap(&arr[i], &arr[j]);
        }
    }
    swap(&arr[i + 1], &arr[high]);
    return (i + 1);
}

// Função para implementar o QuickSort
void quickSort(int arr[], int low, int high) {
    if (low < high) {
        int pi = partition(arr, low, high);
        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}

// Função para imprimir o array
void printArray(int arr[], int size) {
    for (int i = 0; i < size; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");
}

// Função para preencher o array com números aleatórios
void fillArray(int arr[], int size) {
    for (int i = 0; i < size; i++) {
        arr[i] = rand() % 1000; // Preenche com números aleatórios de 0 a 999
    }
}

// Função para medir o tempo médio de ordenação
double measureSortTime(int size, int numTrials) {
    int* arr = (int*)malloc(size * sizeof(int));
    clock_t start, end;
    double totalTime = 0;

    for (int i = 0; i < numTrials; i++) {
        fillArray(arr, size);
        start = clock();
        quickSort(arr, 0, size - 1);
        end = clock();
        totalTime += (double)(end - start) / CLOCKS_PER_SEC;
    }

    free(arr);
    return totalTime / numTrials;
}

int main() {
    int sizes[] = {10, 100, 1000, 10000, 100000};
    int numSizes = sizeof(sizes) / sizeof(sizes[0]);
    int numTrials = 10;

    for (int i = 0; i < numSizes; i++) {
        double avgTime = measureSortTime(sizes[i], numTrials);
        printf("Tamanho: %d, Tempo médio: %f segundos\n", sizes[i], avgTime);
    }

    return 0;
}

File: Proj4a/Questão 4/Quick-sort/quick-sort.java
import java.util.Arrays;
import java.util.Random;

public class QuickSortBenchmark {

    public static void main(String[] args) {
        int[] sizes = {10, 100, 1000, 10000}; // Tamanhos das listas a serem testadas

        // Executando para cada tamanho de lista
        for (int size : sizes) {
            System.out.println("Tamanho da lista: " + size);
            double totalTime = 0; // Para armazenar o tempo total gasto nas ordenações

            // Executando múltiplas vezes para obter um tempo médio
            int numTests = 10; // Número de vezes que cada tamanho de lista será testado
            for (int i = 0; i < numTests; i++) {
                int[] arr = generateRandomArray(size); // Gerando lista aleatória de tamanho 'size'

                // Clonando a lista para manter a original intacta para cada teste
                int[] arrCopy = arr.clone();

                long startTime = System.nanoTime();
                QuickSort ob = new QuickSort();
                ob.sort(arrCopy, 0, arrCopy.length - 1);
                long endTime = System.nanoTime();

                double duration = (endTime - startTime) / 1e6; // Convertendo para milissegundos
                System.out.println("Execução " + (i + 1) + ": " + duration + " ms");
                totalTime += duration;
            }

            double averageTime = totalTime / numTests;
            System.out.println("Tempo médio: " + averageTime + " ms\n");
        }
    }

    // Método para gerar um array de inteiros aleatórios
    private static int[] generateRandomArray(int size) {
        Random random = new Random();
        int[] arr = new int[size];
        for (int i = 0; i < size; i++) {
            arr[i] = random.nextInt(); // Gerando números inteiros aleatórios
        }
        return arr;
    }

    // Classe Quicksort implementada
    static class QuickSort {
        int partition(int arr[], int low, int high) {
            int pivot = arr[high];
            int i = (low - 1);
            for (int j = low; j < high; j++) {
                if (arr[j] < pivot) {
                    i++;
                    int temp = arr[i];
                    arr[i] = arr[j];
                    arr[j] = temp;
                }
            }
            int temp = arr[i + 1];
            arr[i + 1] = arr[high];
            arr[high] = temp;
            return i + 1;
        }

        void sort(int arr[], int low, int high) {
            if (low < high) {
                int pi = partition(arr, low, high);
                sort(arr, low, pi - 1);
                sort(arr, pi + 1, high);
            }
        }
    }
}
